GH-249 Added properties for handling backpressure in KplMessageHandler

siddharthjain210 opened the pull request:
Added 2 properties for handling backpressure. 
Added test cases for KPL Message Handler

artembilan commented on src/test/java/org/springframework/integration/aws/outbound/KplMessageHandlerTest.java:
```diff
@@ -0,0 +1,142 @@
+/**
+ * @author Siddharth Jain
+ *
+ * @since 4.0
```
I think we will include this into the current `3.0.x` generation: nothing really new to postpone it to the next major.
Therefore `@since 3.0.9`

siddharthjain210 replied:
fixed.

artembilan commented on src/test/java/org/springframework/integration/aws/outbound/KplMessageHandlerTest.java:
```diff
@@ -0,0 +1,142 @@
+ */
+@SpringJUnitConfig
+@DirtiesContext
+public class KplMessageHandlerTest {
```
We prefer to name tests classes like `..Tests`.

siddharthjain210 replied:
Ack, renamed.

artembilan commented on src/main/java/org/springframework/integration/aws/outbound/KplMessageHandler.java:
```diff
@@ -99,6 +100,10 @@ public class KplMessageHandler extends AbstractAwsMessageHandler<Void> implement
 
 	private volatile ScheduledFuture<?> flushFuture;
 
+	private long maxRecordsInFlight = 0;
```
Can we name this instead like `maxInFlightRecords`?

siddharthjain210 replied:
renamed.

artembilan commented on src/main/java/org/springframework/integration/aws/outbound/KplMessageHandler.java:
```diff
@@ -99,6 +100,10 @@ public class KplMessageHandler extends AbstractAwsMessageHandler<Void> implement
 
+	private long maxRecordsInFlight = 0;
+
+	private int maxRecordInFlightsSleepDurationInMillis = 100;
```
I don't think there is a reason to mention in the property name that it is "in millis".
Just `maxInFlightRecordsDuration`.
The rest we would just mention in the respective JavaDocs.

siddharthjain210 replied:
Agree. Fixed.

artembilan commented on src/main/java/org/springframework/integration/aws/outbound/KplMessageHandler.java:
```diff
@@ -115,8 +120,29 @@ public void setConverter(Converter<Object, byte[]> converter) {
+	 *
+	 * @param maxRecordsInFlight Defaulted to 0. Value of 0 indicates that Backpressure handling is not enabled. Specify a positive value to enable back pressure.
+	 */
+	public void setMaxOutstandingRecordsInFlight(long maxRecordsInFlight) {
```
Name it like `setMaxInFlightRecords`.
The method JavaDoc must come without blank lines.
We also prefer length of code lines not more than `120` symbols.
Also: `@since 3.0.9`

siddharthjain210 replied:
fixed.

artembilan commented on src/main/java/org/springframework/integration/aws/outbound/KplMessageHandler.java:
```diff
@@ -115,8 +120,29 @@ public void setConverter(Converter<Object, byte[]> converter) {
+	 *
+	 * @param maxRecordInFlightsSleepDurationInMillis Default is 100ms.
+	 */
+	public void setMaxRecordInFlightsSleepDurationInMillis(int maxRecordInFlightsSleepDurationInMillis) {
```
DITTO here: `setMaxInFlightRecordsDuration`.
With respective code style for length and `@since 3.0.9`

artembilan reviewed (changes requested):
This looks promising.
Please, consider my review.
Thanks

siddharthjain210 commented on src/main/java/org/springframework/integration/aws/outbound/KplMessageHandler.java:
```diff
@@ -368,6 +394,16 @@ private void setGlueSchemaIntoUserRecordIfAny(UserRecord userRecord, Message<?>
 	private CompletableFuture<UserRecordResponse> handleUserRecord(UserRecord userRecord) {
+		if (this.maxRecordsInFlight != -1 && this.kinesisProducer.getOutstandingRecordsCount() > this.maxRecordsInFlight) {
+			try {
+				Thread.sleep(this.maxRecordInFlightsSleepDurationInMillis);
```
Can you review this, how does it sound
```
    private void waitForCapacityInUserRecordsBuffer() {
        var kplOutstandingRecordsCount = this.kinesisProducer.getOutstandingRecordsCount();
        var attempts = 1;
        var sleepDuration = this.maxInFlightRecordsBackoffDuration;
        while (kplOutstandingRecordsCount >= this.maxInFlightRecords && 
                attempts <= this.maxInFlightRecordsBackoffMaxAttempts) {
            try {
                logger.info("Buffer size: {} has reached the max records limit of {}. Attempt# {}".formatted(
                        kplOutstandingRecordsCount, this.maxInFlightRecords));
                logger.info("Buffer sleeping for {} ms".formatted(
                        this.maxInFlightRecordsBackoffDuration));
                Thread.sleep(this.maxInFlightRecordsBackoffDuration);
            } catch (InterruptedException ex) {
                Thread.currentThread().interrupt();
            } finally {
                sleepDuration = sleepDuration * this.maxInFlightRecordsBackoffRate;
                attempts++;
                kplOutstandingRecordsCount = this.kinesisProducer.getOutstandingRecordsCount();
            }
        }

        if (kplOutstandingRecordsCount < this.maxInFlightRecords) {
            logger.info("Buffer cleared on number of attempts: {}".formatted(attempts));
            return;
        }

        if (attempts > this.maxInFlightRecordsBackoffMaxAttempts) {
            logger.warn("Buffer not cleared after maximum {} number of attempts & {} sleepDuration".formatted(attempts,
                    sleepDuration));
        }
    }
```

siddharthjain210 replied:
Thinking about oldestrecordtimeinmillis, can also lead to back pressure. There can be too many records on a single millisecond causing a pressure on kpl daemon process. The load pattern would determine how is the distribution of record on each millisecond. 

One can find the optimal value for the max in flight record count by doing a stress test.

siddharthjain210 commented:
@artembilan : I have pushed the changes as requested, please take a look at the test cases to understand the behaviour.

artembilan commented on src/main/java/org/springframework/integration/aws/outbound/KplMessageHandler.java:
```diff
@@ -368,11 +429,50 @@ private void setGlueSchemaIntoUserRecordIfAny(UserRecord userRecord, Message<?>
+	private void waitForCapacityInUserRecordsBuffer() {
+		var kplOutstandingRecordsCount = this.kinesisProducer.getOutstandingRecordsCount();
+		var attempts = 1;
+		var sleepDuration = this.maxInFlightRecordsBackoffDuration;
```
Doesn't look like this variable is used somehow in the logic.

siddharthjain210 replied:
Its a great catch. Thank you @artembilan.

siddharthjain210 commented on src/test/java/org/springframework/integration/aws/outbound/KplMessageHandlerTests.java:
```diff
@@ -0,0 +1,219 @@
+			assertThat(ex).isNotNull();
+			assertThat(ex.getCause()).isNotNull();
+			assertThat(ex.getCause().getClass()).isEqualTo(RuntimeException.class);
+			assertThat(ex.getCause().getClass()).isEqualTo(RuntimeException.class);
```
Thanks, Resolved.

siddharthjain210 commented on src/test/java/org/springframework/integration/aws/outbound/KplMessageHandlerTests.java:
```diff
@@ -0,0 +1,219 @@
+			KplMessageHandler kplMessageHandler = new KplMessageHandler(kinesisProducer);
+			kplMessageHandler.setAsync(true);
+			kplMessageHandler.setStream("foo");
+			kplMessageHandler.setEmbeddedHeadersMapper(new EmbeddedJsonHeadersMessageMapper("foo"));
```
Initially there were no tests for KPL Message Handler. This could have been handy for the upcoming tests.  
Agreed, removed the references to keep only the contextual information.

siddharthjain210 commented on src/main/java/org/springframework/integration/aws/outbound/KplMessageHandler.java:
```diff
@@ -115,8 +124,60 @@ public void setConverter(Converter<Object, byte[]> converter) {
 	}
 
+	/**
+	 * When in KPL mode, the setting allows handling backpressure on the KPL native process.
```
Makes sense. Corrected.

siddharthjain210 commented on src/main/java/org/springframework/integration/aws/outbound/KplMessageHandler.java:
```diff
@@ -115,8 +124,60 @@ public void setConverter(Converter<Object, byte[]> converter) {
+	 * When in KPL mode, the setting allows handling backpressure on the KPL native process.
+	 * Setting this value would enable a sleep on the KPL Thread for the specified number of milliseconds defined in
+	 * maxRecordInFlightsSleepDurationInMillis.
+	 *
```
Removed the blank lines. Cleared the references.

siddharthjain210 commented on src/checkstyle/checkstyle.xml:
```diff
@@ -177,6 +177,10 @@
+
+<!--	Enable this in the future to restrict a Line Length Check.-->
+<!--	<module name="LineLength">-->
+<!--		<property name="max" value="120"/>-->
```
Done.

siddharthjain210 commented on src/main/java/org/springframework/integration/aws/outbound/KplMessageHandler.java:
```diff
@@ -115,8 +124,60 @@ public void setConverter(Converter<Object, byte[]> converter) {
+	 * Specify a positive value to enable back pressure.
+	 * @since 3.0.9
+	 */
+	public void setMaxRecordsInFlight(long maxRecordsInFlight) {
```
i can see that the property, setter are consistent. Anything i am missing?
```
private long maxInFlightRecords = 0;
public void setMaxRecordsInFlight(long maxRecordsInFlight) 
```

siddharthjain210 replied:
I missed correcting this. Thank you.

siddharthjain210 commented on src/main/java/org/springframework/integration/aws/outbound/KplMessageHandler.java:
```diff
@@ -115,8 +124,60 @@ public void setConverter(Converter<Object, byte[]> converter) {
+
+	/**
+	 * The setting allows handling backpressure on the KPL native process.
+	 * Enabled when maxOutstandingRecordsCount is greater than 0.
```
Corrected the java docs, validated them with the generated docs.  Addressed consistency related issues in the javadoc generated.

siddharthjain210 commented on src/main/java/org/springframework/integration/aws/outbound/KplMessageHandler.java:
```diff
@@ -115,8 +124,60 @@ public void setConverter(Converter<Object, byte[]> converter) {
+	}
+
+	/**
+	 * The setting allows handling backpressure on the KPL native process using exponential retry. On attempts
```
Probably the javadoc is causing a confusion as per your comment.  I have removed the mentions "KPL Native Process" from the java docs. 
Please let me know if you think otherwise or my understanding is wrong.

siddharthjain210 commented on src/main/java/org/springframework/integration/aws/outbound/KplMessageHandler.java:
```diff
@@ -368,11 +429,50 @@ private void setGlueSchemaIntoUserRecordIfAny(UserRecord userRecord, Message<?>
 				.thenApply(UserRecordResponse::new);
 	}
 
+	private void waitForCapacityInUserRecordsBuffer() {
```
Its a good thought, sharing my opinions from references with the spring cloud stream kinesis binder. What you are suggesting,  can be very helpful if any consumer is using the spring integration directly rather than transitively. 

1. Any consumer using the spring cloud stream kinesis binder - would require ```KinesisProducer``` exposed as a bean that can be used.  The binder internally constructs ```KinesisProducer```  rather than referencing a bean. 
2.  This can still be baked into the ```KPLMessageHandler``` given, all the consumers of the handler would be replicating the same logic when needed.  As a consumer, it can be a great thought, if the internals are intelligent. 
3. For accomodating the backpressure if we need to use the Resilience Advices - we dont need this change. 

Please suggest the next steps.

artembilan replied:
Well, technically Spring Integration is fully used by Spring Cloud Stream transitively. 
In most cased we indeed don't have direct access to those `MessageHandler` instances used in binders.
However we still can improve those binders to expose some hooks or configuration properties.
In the end it won't matter of end-user if those properties are mapped to the `KPLMessageHandler` or some `RateLimiterRequestHandlerAdvice` and `RequestHandlerRetryAdvice` combination.

I'm not sure why would one need `KinesisProducer` instance from binder, if the logic is going to be covered by respective exposed configuration properties for the binder producer.

I feel that something like `maxInFlightRecords` and some `KplBackPressureException` should be enough for this component. The rest of the logic could be handled by the `RequestHandlerRetryAdvice`.

siddharthjain210 replied:
Fixed, as discussed. 
Changed the handling to a simpler condition. So that the users can use the Spring advice as needed.

siddharthjain210 commented on src/test/java/org/springframework/integration/aws/outbound/KplMessageHandlerTests.java:
```diff
@@ -0,0 +1,219 @@
+
+		@Bean
+		public KinesisProducer kinesisProducer() {
+			return mock(KinesisProducer.class);
```
Referring the mockito javadocs, I couldnt find what you are referring, can you please share.
https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html

![image](https://github.com/user-attachments/assets/3d57926a-4939-488c-859f-3751ead07327)

I have also tried the @Mock, @InjectMock, but apparently its not working with @ServiceActuator somehow.

artembilan replied:
See the last method with `reified` vararg. That's where we can use it without any arguments.
And its Javadoc explains what is going on.

siddharthjain210 replied:
Fixed.

siddharthjain210 commented:
@artembilan : I have worked on the most of the tasks, but not clear on the the 2, which are left open.

siddharthjain210 commented:
understood, and makes sense.  Having a ```KPLBackpressureHandlerAdvice``` extending the spring retry advice is good.  I can 
extract the code from the current ```KPLMessageHandler``` into ```KPLBackpressureHandlerAdvice```.
```
public class KPLBackpressureHandlerAdvice extends RequestHandlerRetryAdvice {
}
``` 


The advices are configured in ```AbstractReplyProducingMessageHandler```,
```
public abstract class AbstractReplyProducingMessageHandler extends AbstractMessageProducingHandler
		implements BeanClassLoaderAware {

	private final List<Advice> adviceChain = new LinkedList<>();
```

but the current outbound adapters are extending ```AbstractAwsMessageHandler``` which internally extends ```AbstractMessageProducingHandler```.  AbstractMessageProducingHandler doesn't have any advices configured. 
```

public abstract class AbstractAwsMessageHandler<H> extends AbstractMessageProducingHandler {
}
```

If we can change the ```AbstractAwsMessageHandler`` to extend ```AbstractReplyProducingMessageHandler``` from ```AbstractMessageProducingHandler```, it would allow users to define Advice Chain on their own.   Otherwise, we can add the support for advice in ```AbstractAwsMessageHandler```.  

@artembilan : Please advice the next steps and if the direction I am heading into seems correct.

artembilan commented:
I might not be clear, but I didn't ask for any custom `Advice`.
That use-case specific `KPLBackpressureHandlerAdvice` would not make sense outside of the `KPLMessageHandler`.

My point is to not have any opinion, but rather just fails with an error when condition has not met yet according to the configuration.
From here it might even better sound like `backPressureThreshold` instead of `maxInFlightRecords`.
Then we just throw that specific `KplBackPressureException` - and everything else is up to end-user how to handle such a situation.
The `RequestHandlerRetryAdvice` would call the handler for the same message again and again.
The `RequestHandlerCircuitBreakerAdvice` would not let the message to pass down to the handler if it is in the `open` state.
You may combine them as well.

The advice is applied from the endpoint. 
See more info in docs https://docs.spring.io/spring-integration/reference/handler-advice.html.
We don't need to do anything with the `AbstractAwsMessageHandler` since you simply can proxy the whole handler or rely on the logic from Spring Integration.

siddharthjain210 commented:
Sorry, if i understood correctly this time. 

If the threshold has breached, we throw a KPLbackPressureException. 
The user of KPLMessageHandler can configure any number of advice [retry, timeout, circuit breaker, combinations] as required whenever the above exception is thrown. 

We don't really need to define and write any custom resilience pattern in the KPL message handler. A plain condition would be simpler.

artembilan commented:
> A plain condition would be simpler.

Yes, that's my idea.
I might miss something, but feels like everywhere else.
We fail with some exception because we cannot handle yet.
"Back pressure" in your terms.
So, that is up to user of this service how to deal with such an error.
And lucky we have all the possible tools with those out-of-the-box advices.

siddharthjain210 commented:
@artembilan: I have done the requested changes. Can you please review.
1. Simplified the condition, removed the retry and exp backoff handling as discussed above. 
2. Introduced a ```KPLBackpressureException```.
3. Rename ```maxRecordsInFlight``` to ```backpressureThreshold```.
4. Refactored Tests.

artembilan commented on src/test/java/org/springframework/integration/aws/outbound/KplMessageHandlerTests.java:
```diff
@@ -0,0 +1,170 @@
+			this.kinesisSendChannel.send(message);
+		}
+		catch (Exception ex) {
+			assertThat(ex).isNotNull();
```
See `assertThatExceptionOfType()` instead.

artembilan commented on src/test/java/org/springframework/integration/aws/outbound/KplMessageHandlerTests.java:
```diff
@@ -0,0 +1,170 @@
+		}
+
+		@Bean
+		@ServiceActivator(inputChannel = "kinesisSendChannel")
```
Any chances to have some test where we indeed use one of our advice , e.g. the mentioned retry one?

artembilan reviewed (changes requested):
Thank you for the update !
Please, don’t comment like `Fixed`, `Done` etc. if you agree with me and going to push. I definitely will it is fixed in your next commit(s). Otherwise it feels like a spam in my mail box and contrpriductive.

artembilan commented on src/main/java/org/springframework/integration/aws/outbound/KplMessageHandler.java:
```diff
@@ -115,9 +122,26 @@ public void setConverter(Converter<Object, byte[]> converter) {
+	 */
+	public void setBackPressureThreshold(long backPressureThreshold) {
+		Assert.isTrue(backPressureThreshold >= 0,
+				"'backPressureThreshold must be greater than equal to 0.");
```
What if I would like to disable it?
Looks our logic allows us to use `0`, and this property might be set by some conditional config where value could be resolved to `0`.

siddharthjain210 replied:
0 being allowed value, the setter should allow it too. There could be cases, where a user would want to disable after enabling it. Even if 0 is allowed, One wouldn't be able to disable it back, if we have only ```Assert`greater than 0```. Hence I had relaxed it back. Negative value wouldnt mean anything, which is still covered. 

One simple occurrence was in test cases failure in the same suite, the state for the threshold was persisted due to other test cases running in random order.

artembilan replied:
Sorry, missed a `>=` operator.
So, all good here, unless we would fix the exception message with an `or` between `than` and `equal`.

artembilan commented on src/main/java/org/springframework/integration/aws/outbound/KplMessageHandler.java:
```diff
@@ -172,9 +196,12 @@ public void setSequenceNumberExpression(Expression sequenceNumberExpression) {
 	 * into the record data.
+	 *
 	 * @since 2.0
+	 *
```
Please, revert all the changes in Javadocs. 
We prefer to not have blank lines in method Javadocs.

artembilan commented on src/main/java/org/springframework/integration/aws/outbound/KplMessageHandler.java:
```diff
@@ -403,7 +451,8 @@ private PutRecordRequest buildPutRecordRequest(Message<?> message) {
 			}
-			Assert.state(partitionKey != null, "'partitionKey' must not be null for sending a Kinesis record. "
+			Assert.state(partitionKey != null,
+					"'partitionKey' must not be null for sending a Kinesis record. "
```
This change doesn't look reasonable.

siddharthjain210 replied:
The line length was crossing beyond 120, at-least I could fix them to reduce some of the debt.

artembilan reviewed (changes requested):
Almost there.
Thank you for the update!

artembilan commented on src/main/java/org/springframework/integration/aws/outbound/KplMessageHandler.java:
```diff
@@ -63,11 +64,16 @@
- * Library {@code putRecord(s)}.
+ * The {@link AbstractMessageHandler} implementation for the Amazon Kinesis Producer Library {@code putRecord(s)}.
+ * <p>
+ * {@link KplBackpressureException} is thrown when backpressure handling is enabled and buffer is at max capacity.
```
I guess English way is to have `The` in the beginning of this sentence.

artembilan commented on src/main/java/org/springframework/integration/aws/outbound/KplMessageHandler.java:
```diff
@@ -63,11 +64,16 @@
+ * The {@link AbstractMessageHandler} implementation for the Amazon Kinesis Producer Library {@code putRecord(s)}.
+ * <p>
+ * {@link KplBackpressureException} is thrown when backpressure handling is enabled and buffer is at max capacity.
+ * The exception can be handled with
```
`This exception`?
I mean sounds better and have a connection with the previous sentence.

artembilan commented on src/main/java/org/springframework/integration/aws/outbound/KplMessageHandler.java:
```diff
@@ -115,9 +122,26 @@ public void setConverter(Converter<Object, byte[]> converter) {
+	 * On number of records in flight exceeding the threshold, {@link KplBackpressureException} would be thrown.
+	 * If Backpressure handling is enabled, {@link KplBackpressureException} must be handled.
+	 *
+	 * @param backPressureThreshold Defaulted to 0. Set a value greater than 0 to enable backpressure handling.
```
I meant remove ` Defaulted to 0`.
Doesn't sounds good and covered with the next sentence.

artembilan commented on src/main/java/org/springframework/integration/aws/support/KplBackpressureException.java:
```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2017-2024 the original author or authors.
```
This class was not created in `2017`.
Or better to say: committed to code base.
So, I prefer to start from the actual year when it is committed.
Therefore in our case, since today is the last day of 2024, I'd suggest to change this copyright just to `2025` or `2025-2025`. Just because we might not merge this today.
And, please, change the copyright in other affected classes.

siddharthjain210 replied:
Please confirm if the rest of the related classes, not just the changed classes have to be updated as well. For now, I have updated only the changed classes.

artembilan commented on src/main/java/org/springframework/integration/aws/support/KplBackpressureException.java:
```diff
@@ -0,0 +1,47 @@
+
+/**
+ * An exception triggered from {@link org.springframework.integration.aws.outbound.KplMessageHandler} while sending
+ * records to kinesis when maximum number of records in flight exceeds the backpressure threshold.
```
I believe has to be capitalized - `Kinesis`

artembilan commented on src/test/java/org/springframework/integration/aws/outbound/KplMessageHandlerTests.java:
```diff
@@ -0,0 +1,184 @@
+				.isThrownBy(() -> this.kinesisSendChannel.send(message))
+				.withCauseInstanceOf(MessageHandlingException.class)
+				.withRootCauseExactlyInstanceOf(KplBackpressureException.class)
+				.withStackTraceContaining("Cannot send record to kinesis since buffer is at max capacity.");
```
Please, fix to `Kinesis`.

artembilan reviewed (approved):
LGTM.
Pulling locally for final review and merge.

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration-aws/commit/b5f6c391834245fc4bf94e11b272fe480ddfdac9 after some minor clean up.
Thank you for contribution; looking forward for more!
